## Input and Premise
- input is a block of text with `@` representing a roll of paper and `.` representing nothing. Consider this block as a grid.
- a roll of paper can be accessed by forklift if there are `fewer than four` rolls in the adjacent 8 cells.
- we need to tell the elves how many rolls are forklift accessible
  
## Thoughts
- first need to process text block into 2D array
- need helper function to count adjacent rolls
- translate text to array, check every cell against count functiom, add if <4


In [1]:
def countAdjacent(grid, row, col):
    adjacentCount = 0
    for r in range(row-1, row+1 +1):
        for c in range(col-1, col+1 +1):
            if r == row and c == col:
                continue
            if r < 0 or r >= len(grid) or c < 0 or c >= len(grid[0]):
                continue
            if grid[r][c] == '@':
                adjacentCount += 1
    
    return adjacentCount

In [2]:
infile = open("input.txt", "r")
grid = []
for line in infile:
    line = line.strip()
    grid.append( [c for c in line] )

totalRollsAccessible = 0
for row in range(len(grid)):
    for col in range(len(grid[0])):
        if grid[row][col] == '@':
            if countAdjacent(grid, row, col) < 4:
                totalRollsAccessible += 1

print(f"Total Rolls Accessible: {totalRollsAccessible}")

Total Rolls Accessible: 1416


## Part 2
- same grid mechanic, but now we can remove an accessible roll and get access to more (by same check)
- how many rolls can be removed in this iterative version?

In [3]:
# going to keep the same input process
infile = open("input.txt", "r")
grid = []
for line in infile:
    line = line.strip()
    grid.append( [c for c in line] )

numRemovedRound = -1
totalRollsRemoved = 0
roundNum = 0
# going to keep passing through the grid until no more can be removed
# instantly remove a roll instead of doing it in rounds like example
while numRemovedRound != 0:
    numRemovedRound = 0
    for row in range(len(grid)):
        for col in range(len(grid[0])):
            if grid[row][col] == '@':
                if countAdjacent(grid, row, col) < 4:
                    grid[row][col] = '.'
                    numRemovedRound += 1
                    totalRollsRemoved += 1
    print(f"Removed {numRemovedRound} rolls on sweep {roundNum} of grid")
    roundNum += 1

print(f"Total Rolls Removed: {totalRollsRemoved}")

Removed 2946 rolls on sweep 0 of grid
Removed 1433 rolls on sweep 1 of grid
Removed 1095 rolls on sweep 2 of grid
Removed 761 rolls on sweep 3 of grid
Removed 593 rolls on sweep 4 of grid
Removed 433 rolls on sweep 5 of grid
Removed 325 rolls on sweep 6 of grid
Removed 266 rolls on sweep 7 of grid
Removed 205 rolls on sweep 8 of grid
Removed 184 rolls on sweep 9 of grid
Removed 134 rolls on sweep 10 of grid
Removed 116 rolls on sweep 11 of grid
Removed 109 rolls on sweep 12 of grid
Removed 80 rolls on sweep 13 of grid
Removed 77 rolls on sweep 14 of grid
Removed 47 rolls on sweep 15 of grid
Removed 39 rolls on sweep 16 of grid
Removed 34 rolls on sweep 17 of grid
Removed 27 rolls on sweep 18 of grid
Removed 29 rolls on sweep 19 of grid
Removed 19 rolls on sweep 20 of grid
Removed 17 rolls on sweep 21 of grid
Removed 13 rolls on sweep 22 of grid
Removed 13 rolls on sweep 23 of grid
Removed 9 rolls on sweep 24 of grid
Removed 7 rolls on sweep 25 of grid
Removed 6 rolls on sweep 26 of gri

# Done!!

## Part 1
- Answer: 1416
- very straightforward, adjacency check was trickiest with walls, but alright

## Part 2
- Answer: 9086
- also very straightforward with this approach, very likely there is a more complicated but more efficient way. It still only took 0.2s in this notebook.